pkg/apis/apps/v1alpha1/staticpod_types.go

```diff
@@ -0,0 +1,146 @@
+type StaticPodUpgradeStrategyType string
+
+const (
+	AutoStaticPodUpgradeStrategyType StaticPodUpgradeStrategyType = "auto"
+	OTAStaticPodUpgradeStrategyType  StaticPodUpgradeStrategyType = "ota"
+)
+
+// StaticPodSpec defines the desired state of StaticPod
+type StaticPodSpec struct {
+	// StaticPodManifest indicates the Static Pod desired to be upgraded. The corresponding
+	// manifest file name is `StaticPodManifest.yaml`.
+	StaticPodManifest string `json:"staticPodManifest,omitempty"`
+
+	// An upgrade strategy to replace existing static pods with new ones.
+	UpgradeStrategy StaticPodUpgradeStrategy `json:"upgradeStrategy,omitempty"`
+
+	// The number of old history to retain to allow rollback.
+	// Defaults to 10.
+	// +optional
+	RevisionHistoryLimit *int32 `json:"revisionHistoryLimit,omitempty"`
+
+	// An object that describes the desired upgrade static pod.
+	// +optional
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +kubebuilder:validation:Schemaless
+	Template corev1.PodTemplateSpec `json:"template,omitempty"`
+}
+
+type StaticPodConditionType string
+
+const (
+	// StaticPodUpgradeSuccess means static pods on all nodes have been upgraded to the latest version
+	StaticPodUpgradeSuccess StaticPodConditionType = "UpgradeSuccess"
+
+	// StaticPodUpgradeExecuting means static pods upgrade task is in progress
+	StaticPodUpgradeExecuting StaticPodConditionType = "Upgrading"
+
+	// StaticPodUpgradeFailed means that exist pods failed to upgrade during the upgrade process
+	StaticPodUpgradeFailed StaticPodConditionType = "UpgradeFailed"
+)
+
+// StaticPodCondition describes the state of a StaticPodCondition at a certain point.
+type StaticPodCondition struct {
+	// Type of StaticPod condition.
+	Type StaticPodConditionType `json:"type,omitempty"`
+
+	// Status of the condition, one of True, False, Unknown.
+	Status corev1.ConditionStatus `json:"status,omitempty"`
+
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+	// The reason for the condition's last transition.
+	Reason string `json:"reason,omitempty"`
+
+	// A human-readable message indicating details about the transition.
+	Message string `json:"message,omitempty"`
+}
+
+// StaticPodStatus defines the observed state of StaticPod
+type StaticPodStatus struct {
+	// The total number of nodes that are running the static pod.
+	TotalNumber int32 `json:"totalNumber"`
+
+	// The number of nodes that are running static pods which need to be upgraded.
+	DesiredNumber int32 `json:"desiredNumber"`
+
+	// The number of nodes that are running updated static pod.
+	UpgradedNumber int32 `json:"upgradedNumber"`
+
+	// The most recent generation observed by the static pod controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration"`
+
+	// Represents the latest available observations of StaticPod's current state.
+	// +optional
+	Conditions []StaticPodCondition `json:"conditions"`
```

Comment:
If StaticPod is defined as a workload like DaemonSet, Conditions field is not needed. so how about remove `Conditions`?

Reply:
done